Agent fails to get active or latest replica set

Happended once locally with minikube.

Agent paniced:
```
time="2018-02-01T12:12:57Z" level=info msg="kubernetes: targeting api server https://10.0.0.1:443"
time="2018-02-01T12:12:57Z" level=info msg="Updating self from https://get.dev.weave.works/k8s/agent.yaml"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xd85485]

goroutine 21 [running]:
github.com/weaveworks/launcher/vendor/k8s.io/api/extensions/v1beta1.(*ReplicaSet).GetAnnotations(0x0, 0x0)
        <autogenerated>:1 +0x5
github.com/weaveworks/launcher/vendor/k8s.io/kubernetes/pkg/controller/deployment/util.Revision(0x1a11380, 0x0, 0x1, 0x1, 0x0)
        /go/src/github.com/weaveworks/launcher/vendor/k8s.io/kubernetes/pkg/controller/deployment/util/deployment_util.go:236 +0x69
github.com/weaveworks/launcher/pkg/k8s.GetDeploymentReplicaSetRevision(0xc4203362d0, 0x12051ed, 0x5, 0x1208ed7, 0xb, 0x0, 0x0, 0x0)
        /go/src/github.com/weaveworks/launcher/pkg/k8s/deployments.go:26 +0x29f
main.updateAgents(0x7fff6cedae02, 0x2a, 0xc4200e4100, 0x3b, 0x1204223, 0x3, 0x7fff6cedae37, 0x6, 0xc4203362d0, 0xc42005a480)
        /go/src/github.com/weaveworks/launcher/agent/main.go:61 +0x128
main.main.func1(0x0, 0x0)
        /go/src/github.com/weaveworks/launcher/agent/main.go:166 +0xe0
github.com/weaveworks/launcher/vendor/github.com/oklog/run.(*Group).Run.func1(0xc420050ae0, 0xc420050a20, 0xc42028dfb0)
        /go/src/github.com/weaveworks/launcher/vendor/github.com/oklog/run/group.go:38 +0x27
created by github.com/weaveworks/launcher/vendor/github.com/oklog/run.(*Group).Run
        /go/src/github.com/weaveworks/launcher/vendor/github.com/oklog/run/group.go:37 +0xa8
```

Instead of panic, error instead - PR: https://github.com/weaveworks/launcher/pull/45

--- a/agent/main.go
+++ b/agent/main.go
@@ -58,21 +58,23 @@ func updateAgents(agentPollURL, wcPollURL string, agentCtx agentContext, kubeCli
 	// Self-update
 	log.Info("Updating self from ", agentPollURL)
 
-	initialRevision, err := k8s.GetDeploymentReplicaSetRevision(kubeClient, "weave", "weave-agent")
+	initialRevision, err := k8s.GetLatestDeploymentReplicaSetRevision(kubeClient, "weave", "weave-agent")
 	if err != nil {
 		logError("Failed to fetch latest deployment replicateset revision", err, agentCtx)
 		return
 	}
+	log.Info("Revision before self-update: ", initialRevision)
 	_, err = kubectl.Execute("apply", "-f", agentPollURL)
 	if err != nil {
 		logError("Failed to execute kubectl apply", err, agentCtx)
 		return
 	}
-	updatedRevision, err := k8s.GetDeploymentReplicaSetRevision(kubeClient, "weave", "weave-agent")
+	updatedRevision, err := k8s.GetLatestDeploymentReplicaSetRevision(kubeClient, "weave", "weave-agent")
 	if err != nil {
 		logError("Failed to fetch latest deployment replicateset revision", err, agentCtx)
 		return
 	}
+	log.Info("Revision after self-update: ", updatedRevision)
 
 	// If the agent replica set is updating, we will be killed via SIGTERM.
 	// The agent uses a RollingUpdate strategy, so we are only killed when the

--- a/pkg/k8s/deployments.go
+++ b/pkg/k8s/deployments.go
@@ -8,8 +8,9 @@ import (
 	deploymentutil "k8s.io/kubernetes/pkg/controller/deployment/util"
 )
 
-// GetDeploymentReplicaSetRevision gets the revision of the latest replica set of a deployment
-func GetDeploymentReplicaSetRevision(kubeClient *kubeclient.Clientset, namespace, name string) (int64, error) {
+// GetLatestDeploymentReplicaSetRevision gets the latest revision of replica sets of a deployment
+func GetLatestDeploymentReplicaSetRevision(kubeClient *kubeclient.Clientset, namespace, name string) (int64, error) {
+	// Based on https://github.com/kubernetes/kubernetes/blob/release-1.9/pkg/kubectl/history.go
 	versionedClient := kubeClient.ExtensionsV1beta1()
 
 	deployment, err := versionedClient.Deployments(namespace).Get(name, metav1.GetOptions{})
@@ -17,16 +18,25 @@ func GetDeploymentReplicaSetRevision(kubeClient *kubeclient.Clientset, namespace
 		return 0, fmt.Errorf("failed to retrieve deployment: %s", err)
 	}
 
-	_, oldRs, newRs, err := deploymentutil.GetAllReplicaSets(deployment, versionedClient)
+	_, allOldRSs, newRS, err := deploymentutil.GetAllReplicaSets(deployment, versionedClient)
 	if err != nil {
 		return 0, fmt.Errorf("failed to retrieve deployment replicasets: %s", err)
 	}
-
-	rs := deploymentutil.FindActiveOrLatest(newRs, oldRs)
-	revision, err := deploymentutil.Revision(rs)
-	if err != nil {
-		return 0, fmt.Errorf("failed to retrieve deployment replicaset revision: %s", err)
+	allRSs := allOldRSs
+	if newRS != nil {
+		allRSs = append(allRSs, newRS)
 	}
 
-	return revision, nil
+	var maxRevision int64
+	for _, rs := range allRSs {
+		v, err := deploymentutil.Revision(rs)
+		if err != nil {
+			continue
+		}
+		if v > maxRevision {
+			maxRevision = v
+		}
+	}
+
+	return maxRevision, nil
 }

